### DISCUSSION TITLE
Custom Autocomplete and understanding params

### DISCUSSION THREAD

**Discussion by brsanthu** [eclipse-langium/langium#1682] [Q&A]
We are implementing custom auto-complete implementation for our DSL.

When it asks to provide completions, it passes `context`, `next` and `acceptor` parameters. Do we have some docs what those params means and how we could infer the context?

Additional details. We are implementing SQL like language. When user types `select id from <cursor>`, we want to return list of objects. How can we know user is at context where we expect list of objects to show?

Update1:
We just logged 3 params and observed as below. When our dsl is implementing like this stmt `select id from account`, the `next.property` is returned as `object` but when we expand to support statement like below

```
select id from account
chain
select id from contact
```

the `next.property` is blank. Both instances were observed when we place the cursor at `select id from <cursor>`. Any idea why `next.property` is blank when we add support for `chain` statement as shown above?

**Comment by msujew** [comment]
Hey @brsanthu,

> When it asks to provide completions, it passes context, next and acceptor parameters.

I assume you're talking about methods like [`completionFor`](https://github.com/eclipse-langium/langium/blob/335f9a99b1568829cd1c0f5d3bbe33ca754daf90/packages/langium/src/lsp/completion/completion-provider.ts#L406)? They're not designed to be invoked from adopter code. Usually, most people override either `completionForKeyword` or `completionForCrossReference` to fine tune the completion results for specific language features.

Aside from them, the type definitions and the names of the arguments should be most of what you need. What info are you missing?

> How can we know user is at context where we expect list of objects to show?

That's what the `next` argument is for.

> Any idea why next.property is blank when we add support for chain statement as shown above?

No idea, it could be any reason (literally - the completion engine is fairly complicated).

Have you taken a look at our [SQL implementation in Langium](https://github.com/TypeFox/langium-sql)? Completion support works fairly well there, except for one caveat (see https://github.com/eclipse-langium/langium/issues/1076).

**Reply by durianwaffle** [reply]
Hi @msujew, I am wondering what's the difference in all the contexts in `getCompletion`?
[langium/packages/langium/src/lsp/completion/completion-provider.ts at 3170af97638ec2cc96d47f694f22a1caeb0c6cd1 · eclipse-langium/langium](https://github.com/eclipse-langium/langium/blob/3170af97638ec2cc96d47f694f22a1caeb0c6cd1/packages/langium/src/lsp/completion/completion-provider.ts#L106)

**Can I use the last context?**

![image](https://github.com/user-attachments/assets/0cf20c23-703f-4cd9-b9b2-ba686b9ab18c)

**Reply by msujew** [reply]
The main difference between the different `CompletionContext` objects is the position in the grammar. I.e. which features it attempts to complete. Even at the same position in the text, there are multiple positions in the grammar which might apply.

> Can I use the last context?

You can override the `continueCompletion` method to return `true` to prevent early exit from the loop.
